### DISCUSSION TITLE
How can one use an external optimizer with LightningCLI?

### DISCUSSION THREAD

**Discussion by goncalomcorreia** [Lightning-AI/pytorch-lightning#11016] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
I would like to use [Adafactor](https://huggingface.co/docs/transformers/main_classes/optimizer_schedules#transformers.Adafactor) as my optimizer with LightningCLI. I've tried the method described in the [documentation](https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_cli.html#optimizers-and-learning-rate-schedulers) for custom optimizers but it didn't work. Can anybody tell me how they would train a model with this optimizer using LightningCLI?

**Comment by tchaton** [comment]
Hey @goncalomcorreia,

Did you add the optimizer to the Optimizer registry?

Here is where we pre-register all the optimizers: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/utilities/cli.py#L83

**Comment by goncalomcorreia** [comment, marked as answer]
Hi! I got it to work in the meantime. I added this to the main file where I call CLI:

```
import transformers
from pytorch_lightning.utilities.cli import OPTIMIZER_REGISTRY

@OPTIMIZER_REGISTRY
class Adafactor(transformers.Adafactor):
    def __init__(self, *args: Any, **kwargs: Any) -> None:
        super().__init__(*args, **kwargs)
```

The main issue was in the config file---apparently one needs to write:

```
optimizer:
  class_path: __main__.Adafactor
```

instead of:


```
optimizer:
  class_path: Adafactor
```

Doing the former, I got it to work.

By the way, is there a way to have the optimizer register in a separate file than the one that calls CLI?

**Reply by mauvilsa** [reply]
If you are using a config file there is no need to use the registry or have this subclass. In the question you said that what is described in the documentation didn't work but there is no mention on what exactly you did or what was the problem. If you make the optimizer configurable as mentioned in the documentation in the config file you would just need:

```yaml
optimizer:
  class_path: transformers.Adafactor
```
